### PR TITLE
Fix unpickling/__setstate__ in TransformNode

### PR DESCRIPTION
**Pull request by joergdietrich**
Fix the unpickling error in #4068 and add a test for it.

### CODE REVIEW

File: lib/matplotlib/transforms.py [hunk 1]
```diff
@@ -115,7 +115,7 @@ def __getstate__(self):
     def __setstate__(self, data_dict):
         self.__dict__ = data_dict
         # turn the normal dictionary back into a WeakValueDictionary
-        self._parents = WeakValueDictionary(self._parents)
+        self._parents = WeakValueDictionary(self.__dict__['_parents'])
```
Comment (tacaswell) [inline lib/matplotlib/transforms.py]: Can you explain why this changes anything?

Comment (joergdietrich) [inline lib/matplotlib/transforms.py]: The weak reference is destroyed during pickling, thus self._parents does not exist anymore at this point. __getstate__ however saved it in data_dict['_parents'] from which it is now restored.

Comment (pelson) [inline lib/matplotlib/transforms.py]: I don't think this is doing what you thing it is. Previously we were restoring the _parents from the stored data dictionary, now this is restoring parents from self, which hasn't yet been initialised. Hence, _parents will simply be empty (https://github.com/joergdietrich/matplotlib/blob/%234068/lib/matplotlib/transforms.py#L97).

Comment (pelson) [inline lib/matplotlib/transforms.py]: > I don't think this is doing what you thing it is.

Scrap that. I think I'm talking rubbish. (Friday afternoon!)

Comment (pelson) [inline lib/matplotlib/transforms.py]: I think you've found a bug with the setstate implementation, and I don't think it has previously been tested correctly.

How about simply doing ``data_dict['_parents']`` though. It may be more clear what is going on that way.

Would you also mind adding a test that ``self._parents`` *actually is* accessible after being restored?

Comment (WeatherGod) [inline lib/matplotlib/transforms.py]: I still don't understand... shouldn't `self._parents` be equivalent to
`self.__dict__['_parents']`, since the object dictionary would be where one
accesses attributes of the object?

**Comment by joergdietrich** [comment]
I'm not sure anymore I understand myself what's going on. This PR definitely fixes the exception during unpickling. However, the unpickled figure cannot be `plt.show()`n:

```
/home/joerg/src/matplotlib/lib/python2.7/site-packages/matplotlib-1.5.dev1-py2.7-linux-x86_64.egg/matplotlib/projections/polar.py in get_matrix(self)
    101 
    102     def get_matrix(self):
--> 103         if self._invalid:
    104             limits_scaled = self._limits.transformed(self._scale_transform)
    105             yscale = limits_scaled.ymax - limits_scaled.ymin

AttributeError: 'PolarAffine' object has no attribute '_invalid'
```

I'll investigate.

**Comment by tacaswell** [comment]
@joergdietrich Any progress on this?

**Comment by joergdietrich** [comment]
No, sorry. Unexpectedly got flooded with other things. This PR does not fix the original problem at all. The exception during unpickling in fact does not occur on my devel branch without this PR. I created a bit of a mess with different environments when trying to fix this. This PR should be closed without merging.

It will be a few weeks before I can look into #4068. I'd be happy if somebody else wants to take a look at it.

**Comment by tacaswell** [comment]
@joergdietrich No worries.  I have great sympathy for both being over-subscribed and python environments fighting back!
